### ISSUE TITLE
GridBox widget

### ISSUE THREAD

**Issue by jasongrout** [jupyter-widgets/ipywidgets#1942]
Now that [CSS grid](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Grid_Layout) is making its way into [browsers](https://caniuse.com/#feat=css-grid), I think it would be great to have (in core widgets) a GridBox widget that places its widgets in a grid using the CSS grid spec.

**Comment by jasongrout** [comment]
Here are some interesting resources from a HN article on CSS grids: https://news.ycombinator.com/item?id=16292094

**Comment by Madhu94** [comment]
I'd like to give this a shot.

**Comment by Madhu94** [comment]
Hi, I had a question - 
Should GridBox inherit from Box? 

There is a lot of shared code that I could reuse. But Box was meant to be a flexbox implementation. I could (1) create an abstract class, say "CSSLayout" (name TBD, but users shouldn't touch this) and move the common functionality into that. Box and GridBox would both subclass this and set ``display`` to either ``grid`` or ``flex``. 

(Note - I was able to get this to work even otherwise though, just by making GridBox inherit from Box and changing display in the child's constructor )

**Comment by jasongrout** [comment]
I think Gridbox should not inherit from Box. The big thing that Box gives you is a children attribute, but that's probably not useful in a 2d layout to have a single list of children.

**Comment by Madhu94** [comment]
I agree that the BoxModel contains just a children attribute, but the BoxView seems to contain some logic that we might end up re-implementing ?

> that's probably not useful in a 2d layout to have a single list of children.
Seems like a good time to finalize the API - did you have a 2D list (list of lists) in mind, as the input ?
I thought it would be a simple list, and since we will expose the grid spec as part of the ``Layout``, the user can set the ``grid-template-columns`` css property to tweak the layout. I've exposed ``grid``, ``grid_area`` and ``grid_gap`` - following the convention of using only shorthand css properties.

**Comment by Madhu94** [comment]
Example : 
``` 
GridBox([w1, w2, w3, w4, w5, w6], layout=Layout(grid='none / repeat(3, 1fr)'))
``` 

This will create 2 rows of 3 columns each

**Comment by jasongrout** [comment]
> Seems like a good time to finalize the API

Good question. I haven't read too deeply into the grid spec, so I'm not up to speed on what would be a natural api.

Layouts are used as a generic mechanism across all widgets, though. I imagine we would have more specific grid-type properties on the widget itself?

**Comment by Madhu94** [comment]
I'm not sure what you mean by grid-specific. Involve the grid properties in the constructor or expose them as attributes ?

Some other API suggestions off the top of my head - 
1. This is more like syntactic sugar over ``Layout``
```
GridBox([w1, w2, w3], {"grid_template_columns": "20% 40% 40%"})
``` 

2.  The user gives us a 2D list and we "guess" how it would be laid out ? 
```
GridBox([[w1, w2,  w3], [w3, w4]])  
```   

3. GridBox will be completely flexible (I do agree that using Layout often is slightly inconvenient). We will build widgets on top of GridBox (l bet VBox and HBox are used more than Grid). 

What would you expect (as a user) from GridBox ? If you don't recollect the CSS Grid spec, your idea of a natural API might be more intuitive :)

**Comment by maartenbreddels** [comment]
What about looking at the wxWidgets Qt API's. I found the wxWidgets sizers to be quite intuitive. Maybe setting it all at once at the contructor is not the way to go, and the details of how the information layout is encoded should be hidden from users.

**Comment by Madhu94** [comment]
I'm not a fan of my suggestions either.:) I just can't think of anything
better than GridBox taking a list of widgets and maybe have the css grid
spec exposed as attributes on GridWidget.

I will try to think of something better and put up a WIP PR.


Do we want to hide the implementation from the user ? For Box, the docs
clearly say that Box exposes the entire FlexBox spec and I thought that was
a great idea, it gives users a lot of power.

**Comment by SylvainCorlay** [comment]
I agree with @Madhu94 about taking a single list of widgets and transparently exposing the CSS grid attributes.

**Comment by martinRenou** [comment]
Hi @Madhu94, I was thinking about working on it today. Did you start already or can I do it?

**Comment by Madhu94** [comment]
Yes, I did start. Did you want to work on this ?

**Comment by Madhu94** [comment]
Thank you so much.

**Comment by SylvainCorlay** [comment]
@Madhu94 don't hesitate to show what you  have so far if you are working on this.

**Comment by Madhu94** [comment]
https://github.com/jupyter-widgets/ipywidgets/pull/2064 I made a WIP PR, let me know if there is something wrong

**Comment by jasongrout** [comment]
Looks like this is fixed by #2107
